nanodbc/nanodbc.h

```diff
@@ -1323,6 +1323,27 @@ class result
     /// \brief Returns true if there are no more results in the current result set.
     bool at_end() const noexcept;
 
+    /// \brief Unbind data buffers all columns in the result set.
+    ///
+    /// Wraps unbind(const std::vector<short>& idx)
+    void unbind();
+
+    /// \brief Unbind data buffers for specific columns in the result set.
+    ///
+    /// When a result is constructed, in order to optimize data retrieval,
+    /// we automatically try to bind buffers, except for columns that contain
+    /// long/blob data types.  This method gives the caller the option to unbind
+    /// specific buffers.  Subsequently, during calls to get(), if there is no
+    /// bound data buffer, we will attempt to retrieve the data using a call
+    /// SQLGetData; this is similar to the route taken for columns hosting long
+    /// or bloby data types.  This is suboptimal from efficiency perspective,
+    /// however may be necessary of the driver we are communicating with does
+    /// not support out-of-order retrieval of long data.
+    ///
+    /// \param idx Vector of zero-based indices of columns we wish to unbind.
+    /// \throws programming_error
+    void unbind(const std::vector<short>& idx);
```

Comment:
I wonder if we could have a nicer interface for this using [C++11 variadic arguments](https://en.cppreference.com/w/cpp/language/variadic_arguments) or perhaps [list initialization](https://en.cppreference.com/w/cpp/language/list_initialization) instead of passing in a vector? Just thinking about how it would typically be used in code. I figure most of the time the user would know which columns they want to have be unbound and it would be an additional step to construct a vector out of those values... tho maybe it's not too bad. I'm not sure. What do you think?

Reply:
> I figure most of the time the user would know, at the time of writing the code, which columns they want to have be unbound

As a user (by way of the R packages odbc, DBI, dplyr and dbplyr) I can tell you I frequently encounter the issue this fix aims to address by executing simple SQL statements like "SELECT * FROM table_with_long data".

Reply:
@nickv9 It seems like in that case that the real problem is the eager auto-binding that nanodbc does is working against you. Maybe it would be nice for that use-case to have a way of telling nanodbc to default to _not_ auto-binding anything, and then you could optionally bind things later or just let nanodbc handle it as necessary? What do you think?

Reply:
@lexicalunit I am not yet familiar technically with the concept of, "binding," a column so I won't answer your question directly right now. From what I've learned the errors other R users and are I are seeing is a result of `SQLGetData` being called before data has been retrieved for all bound (non "long") columns. I did see this comment on the [Microsoft documentation re: long data](https://docs.microsoft.com/en-us/sql/odbc/reference/develop-app/getting-long-data?view=sql-server-ver15):

> An application can actually retrieve any type of data with SQLGetData, not just long data, although only character and binary data can be retrieved in parts.

I do wonder if a practical and simple solution is to just get data for all columns via SQLGetData to completely bypass this ODBC constraint. Some people have noted potential performance concerns. I wonder how tangiable the performance impact would really be in practicality. I'd be willing to benchmark this in my Windows (client and SQL server) environment.

I've done an inventory of this issue in the R community and also my notes to date on debugging [here](https://www.nickvasile.com/2020/05/13/error-invalid-descriptor-index/)

Reply:
> I wonder if we could have a nicer interface for this using [C++11 variadic arguments](https://en.cppreference.com/w/cpp/language/variadic_arguments) or perhaps [list initialization](https://en.cppreference.com/w/cpp/language/list_initialization) instead of passing in a vector?

@lexicalunit : Thanks for the feedback:

*  `results.unbind({2})`, does/should work with the current interface - happy to update the test syntax if you think it makes the use case more legible.

* In terms of how it might be used: the current API is structured as `return_type results::column_some_information(column index)`, where `some_information` might be size, type, etc.  I thought one approach a downstream developer might take is, in pseudo code:
```
idx_to_unbind = std::vector<short>;
for(short i = 0, i<result.columns(); ++i) {
    if result.some_information(i) == some condition
        idx_to_unbind.append(i)
        // or alternatively
       // result.unbind({i});
}
result.unbind(idx_to_unbind)
```
Ultimately, I suspect the information we need to expose is whether there is a buffer bound to a column or not so someone who is interested in unbinding everything past the first unbound column might do:
```
short i = 0;
idx_to_unbind = std::vector<short>;
num_cols = result.columns()
while( i < num_cols)) {
    if !result.is_bound(i) {
        break;
   ++i;
    }
}
idx_to_unbind.resize(num_cols - i);
std::iota(idx_to_unbind.begin(), idx_to_unbind.end(), i);
result.unbind(idx_to_unbind);
```
If this seems reasonable should I add a method `results::is_bound()`?

* RE: optional `auto_bind` versus the ability to unbind specific columns post the greedy bind.  The one challenge I encountered is that `nanodbc` instantiates results in multiple places internally, and the `auto_bind` happens when the result is constructed.   So this parameter, giving the caller the choice to opt out, would need to percolate up to quite a few places (result constructor, statement constructor, free methods that return results).  It caused a bit of churn, but happy to go that route if you think that's better.

Reply:
Hi @lexicalunit 

* I touched up the interface so that it now resembles the remaining column specific methods in `results`.  In particular there is now: `unbind(const string& column_name)`, `unbind(short column)`, as well as a blanket `unbind()` that unbinds all columns.
* I also added the `results::is_bound` method, structured similarly.

Thanks

Reply:
@detule Thank you for the update. I'm reviewing this PR now.